src/libs/opts/opts.c

```diff
@@ -0,0 +1,951 @@
+ *
+ * A short option cannot be set to 'optional', as the only way to give an 'optional' argument is via
+ * the "--opt=value" syntax of long options. If a long option with an 'optional' argument does not
+ * have an argument, its value will be set as if `opt/arg` were 'none'. One difference to getopt_long(3)
+ * is that options with 'optional' arguments can also have short versions. In this case the short version
+ * is treated as if `opt/arg` were set to 'none', i.e. it cannot have an argument.
+ *
+ * If `opt/arg` is set to 'none' the the corresponding key will have the value "1", if the option is
+ * provided. This value can be changed by setting `opt/flagvalue` to something else.
+ *
+ * A key can also be associated with multiple options. To achieve this, simply follow the instructions
+ * above, but replace `opt` with `opt/#XXX` (XXX being the index of the option) in all keynames. Each
+ * option can have a different setting for none/required/optional arguments and flagvalue too.
+ *
+ * In addition to command line options this function also supports environment variables. These are
+ * specified with `env` (or `env/#XXX` if multiple are used).
+ *
+ * While you can specify multiple options (or environment variables) for a single key, only one of them
+ * can be used at the same time. Using two or more options (or variables) that are all linked to the
+ * same key, will result in an error.
+ *
+ * If the key for which the options are defined, has the basename '#', an option can be repeated.
+ * All occurrences will be collected into an array. Environment variables obviously cannot be repeated,
+ * instead a behaviour similar that used for PATH is adopted. On Windows the variable will be split at
+ * each ';' character. On all other systems ':' is used as a separator.
+ *
+ * In case multiple versions (short, long, env-var) of an option are found, the order of precedence is:
+ * <ul>
+ * 	<li> Short options always win. </li>
+ * 	<li> Long options are used if no short option is present. </li>
+ * 	<li> If neither a long nor short option is found, environment variables are considered. </li>
+ * </ul>
+ *
+ * Lastly, all remaining elements of @p argv will be collected into an array. You can access this array
+ * by specifying `args = remaining` on a key with basename '#'. The array will be copied into this key. As is
+ * the case with getopt(3) processing of options will stop if '--' is encountered in @p argv.
+ *
+ * NOTE: Both options and environment variables can only be specified on a single key. This is because
+ * otherwise e.g. the help message ('opt/help', 'env/help') may be ambiguous. If you need to have the
+ * value of one option/environment variable in multiple keys, you may use fallbacks.
+ *
+ * NOTE: Per default option processing DOES NOT stop, when a non-option string is encountered in @p argv.
+ * If you want processing to stop, set the metadata `posixly = 1` on @p errorKey.
```

Comment:
Do we need this feature? If possible we should avoid hard-coded configuration within applications so that users have the same behavior when using different applications.

Reply:
I included it to mimic `POSIXLY_CORRECT` from getopt_long(3). The feature may be useful for an interface like the one `git` uses. General options like `--version` are specified before the command. The first argument in `argv` that is not an option is used as the command. Everything after the command is treated as arguments to the command.

This is a bit harder to implement with `elektraGetOpts`, because you don't automatically get the index of the command, but would still be possible.

Reply:
This can be implemented as later extension. For the `kdb` tool we also need a functionality like this. Would be great if the `kdb` tool also uses your library. And if the OPTIONS sections of the man pages are generated from the --help output. (or a tool directly parsing the specification)

@ulschaef already has some ideas how to extend the specification so that "subcommands" can be supported.

Reply:
> This can be implemented as later extension. 

It is implemented already. The only thing I would (maybe) add to better support "subcommands" is a way to assign the index of the first non-option argument to a key (or maybe even return it directly in some way to avoid conversion to string an back again). Currently you would need to calculate this index from the size of the remaining args array.

The idea would be:
1. call `elektraGetOpts` with the original `argc` and `argv`, in posixly mode.
2. get the index of the first non-option argument
3. call `elektraGetOpts` with changed `argc` and `argv`, so that `argv[0]` is the subcommand.

> And if the OPTIONS sections of the man pages are generated from the --help output.

If you pass `""` as the usage line and `NULL` as prefix to `elektraGetOptsHelpMessage` the result will only be the options list (you might need to trim some whitespace).

Reply:
Yes, I meant support for subcommands as future extension.

> the result will only be the options list

Nice. Maybe we should also have a `kdb` tool which calls elektraGetOptsHelpMessage for a given specification?